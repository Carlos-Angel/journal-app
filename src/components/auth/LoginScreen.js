import React from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import {
  startLoginWithEmailAndPassword,
  startGoogleLogin,
} from '../actions/auth';

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const [{ email, password }, handleInputChange] = useForm({
    email: 'test@test.com',
    password: '123456',
  });

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginWithEmailAndPassword(email, password));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h3 className='auth__title mb-5'>LoginScreen</h3>
      <form onSubmit={handleLogin}>
        <input
          type='email'
          name='email'
          placeholder='email'
          className='auth__input'
          autoComplete='off'
          value={email}
          onChange={handleInputChange}
        />
        <input
          type='password'
          name='password'
          placeholder='password'
          className='auth__input'
          value={password}
          onChange={handleInputChange}
        />
        <button className='btn btn-primary btn-block' type='submit'>
          Login
        </button>
      </form>
      <hr />
      <div className='auth__social-networks'>
        <p>Login with social networks</p>
        <div className='google-btn' onClick={handleGoogleLogin}>
          <div className='google-icon-wrapper'>
            <img
              className='google-icon'
              src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
              alt='google button'
            />
          </div>
          <p className='btn-text'>
            <b>Sign in with google</b>
          </p>
        </div>

        <Link className='mt-5 link' to='/auth/register'>
          Create a new Account
        </Link>
      </div>
    </>
  );
};
