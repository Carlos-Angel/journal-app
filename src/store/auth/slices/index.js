export * from './auth.slice';
export { default as startGoogleSingIn } from './startGoogleSingIn.thunk';
export { default as checkingAuthentication } from './checkingAuthentication.thunk';
export { default as startRegisterUserWithEmailAndPassword } from './startRegisterUserWithEmailAndPassword.thunk';
export { default as startLoginWithEmailAndPassword } from './startLoginWithEmailAndPassword.thunk';
export { default as startLogout } from './startLogout.thunk';
