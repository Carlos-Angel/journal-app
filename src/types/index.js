export const authTypes = {
  login: '[Auth] login',
  logout: '[Auth] logout',
  uiSetError: '[UI] set Error',
  uiRemoveError: '[UI] remove Error',
};
