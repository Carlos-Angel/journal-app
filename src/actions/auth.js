import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
} from 'firebase/auth';
import { googleAuthProvider } from 'app';
import { authTypes } from 'types';

export const startLoginWithEmailAndPassword = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(123, 'pedro'));
    }, 3500);
  };
};

export const startRegister = ({ email, password, name }) => {
  return (dispatch) => {
    const auth = getAuth();
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => console.error(error.message));
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => dispatch(login(user.uid, user.displayName)))
      .catch((error) => console.error(error.message));
  };
};

export const login = (uid, displayName) => {
  return {
    type: authTypes.login,
    payload: {
      uid,
      displayName,
    },
  };
};
