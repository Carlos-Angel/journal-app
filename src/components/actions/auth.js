import { authTypes } from '../types';

export const startLoginWithEmailAndPassword = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(123, 'pedro'));
    }, 3500);
  };
};

export const login = (uuid, displayName) => {
  return {
    type: authTypes.login,
    payload: {
      uuid,
      displayName,
    },
  };
};
