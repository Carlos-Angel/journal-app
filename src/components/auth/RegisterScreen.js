import React from 'react';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from 'actions/ui';
import { useForm } from 'hooks/useForm';

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { msgError } = useSelector((state) => state.ui);

  const initialValuesForm = {
    email: '',
    name: '',
    password: '',
    confirmPassword: '',
  };
  const [{ email, name, password, confirmPassword }, handleInputChange] =
    useForm(initialValuesForm);

  const handleSubmit = (e) => {
    e.preventDefault();
    if (isFormValid()) {
      dispatch(removeError());
    }
  };

  const isFormValid = () => {
    if (validator.isEmpty(name)) {
      dispatch(setError('name is required'));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError('email is not valid'));
      return false;
    } else if (!validator.equals(password, confirmPassword)) {
      dispatch(setError('password should be match each other'));
      return false;
    } else if (!validator.isLength(password, { min: 8 })) {
      dispatch(setError('password should be at least 8 characters'));
      return false;
    }
    return true;
  };

  return (
    <>
      <h3 className='auth__title mb-5'>Register</h3>
      <form onSubmit={handleSubmit}>
        {msgError && <div className='auth__alert-error'>{msgError}</div>}
        <input
          type='text'
          name='name'
          placeholder='Name'
          className='auth__input'
          autoComplete='off'
          value={name}
          onChange={handleInputChange}
        />
        <input
          type='email'
          name='email'
          placeholder='email'
          className='auth__input'
          autoComplete='off'
          value={email}
          onChange={handleInputChange}
        />
        <input
          type='password'
          name='password'
          placeholder='password'
          className='auth__input'
          value={password}
          onChange={handleInputChange}
        />
        <input
          type='password'
          name='confirmPassword'
          placeholder='Confirm password'
          className='auth__input'
          value={confirmPassword}
          onChange={handleInputChange}
        />
        <button className='mb-5 btn btn-primary btn-block' type='submit'>
          Register
        </button>
      </form>
      <Link className='mt-5 link' to='/auth/login'>
        Already registered, sign in?
      </Link>
    </>
  );
};
