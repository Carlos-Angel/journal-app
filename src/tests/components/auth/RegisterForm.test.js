import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { RegisterScreen } from 'components/auth/RegisterScreen';
import { Provider } from 'react-redux';
import { authTypes } from 'types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: {
    loading: false,
    msgError: '',
  },
};

let store = mockStore(initState);

describe('Pruebas del componente <RegisterScreen />', () => {
  beforeEach(() => {
    store.clearActions();
    jest.clearAllMocks();
  });

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    </Provider>,
  );

  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de hacer el dispatch de ña acción respectiva', () => {
    const emailField = wrapper.find('input[name="email"]');
    emailField.simulate('change', {
      target: {
        value: '',
        name: 'name',
      },
    });

    wrapper.find('form').simulate('submit', {
      preventDefault() {},
    });

    const actions = store.getActions();
    expect(actions[0]).toEqual({
      type: authTypes.uiSetError,
      payload: 'name is required',
    });
  });

  test('debe de mostrar el div de alert error', () => {
    const initState = {
      auth: {},
      ui: {
        loading: false,
        msgError: 'name is required',
      },
    };

    let store = mockStore(initState);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>,
    );

    expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
    expect(wrapper.find('.auth__alert-error').text().trim()).toBe(
      'name is required',
    );
  });
});
