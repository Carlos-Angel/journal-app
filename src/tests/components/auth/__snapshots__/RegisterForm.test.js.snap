// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <RegisterScreen /> debe de mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title mb-5"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="mb-5 btn btn-primary btn-block"
      type="submit"
    >
      Register
    </button>
  </form>,
  <a
    className="mt-5 link"
    href="/auth/login"
    onClick={[Function]}
  >
    Already registered, sign in?
  </a>,
]
`;
